prov/verbs/src/verbs_eq.c

```diff
@@ -144,30 +232,293 @@ fi_ibv_eq_cm_getinfo(struct fi_ibv_fabric *fab, struct rdma_cm_event *event,
+
+	*priv_data = (cm_data + 1);
+	*priv_data_len -= sizeof(*cm_data);
+}
+
+static int
+fi_ibv_eq_xrc_connreq_event(struct fi_ibv_eq *eq, struct fi_eq_cm_entry *entry,
+			    const void **priv_data, size_t *priv_datalen)
+{
+	struct fi_ibv_connreq *connreq = container_of(entry->info->handle,
+						struct fi_ibv_connreq, handle);
+	struct fi_ibv_ep *ep;
+	struct fi_ibv_xrc_cm_data cm_data;
+	int ret;
+
+	if (!connreq->xrc.is_reciprocal) {
+		fi_ibv_eq_skip_xrc_cm_data(priv_data, priv_datalen);
+		return FI_SUCCESS;
+	}
+
+	/*
+	 * Reciprocal connections are initiated and handled internally by
+	 * the provider, get the endpoint that issued the original connection
+	 * request.
+	 */
+	ep = fi_ibv_eq_xrc_conn_tag2ep(eq, connreq->xrc.conn_tag);
+	if (!ep) {
+		VERBS_WARN(FI_LOG_FABRIC,
+			   "Reciprocal XRC connection tag not found\n");
+		goto send_reject;
+	}
+	ep->tgt_id = connreq->id;
+	ep->tgt_info = fi_dupinfo(entry->info);
+	ep->tgt_id->context = &ep->util_ep.ep_fid.fid;
+	ep->info->handle = entry->info->handle;
+
+	ret = rdma_migrate_id(ep->tgt_id, ep->eq->channel);
+	if (ret) {
+		VERBS_WARN(FI_LOG_FABRIC, "Could not migrate CM ID\n");
+		goto err;
+	}
+
+	ret = fi_ibv_accept_xrc(ep, FI_IBV_RECIP_CONN, &cm_data,
+				sizeof(cm_data));
+	if (ret) {
+		VERBS_WARN(FI_LOG_FABRIC,
+			   "Reciprocal XRC Accept failed %d\n", ret);
+		goto err;
+	}
+	return -FI_EAGAIN;
+
+err:
+	fi_freeinfo(ep->tgt_info);
+	ep->tgt_info = NULL;
+
+send_reject:
+	if (rdma_reject(connreq->id, *priv_data, *priv_datalen))
+		VERBS_WARN(FI_LOG_FABRIC, "rdma_reject %d\n", -errno);
+	return -FI_EAGAIN;
+}
+
+static int
+fi_ibv_eq_xrc_conn_event(struct fi_ibv_ep *ep, struct rdma_cm_event *cma_event,
+			 struct fi_eq_cm_entry *entry)
+{
+	struct fi_ibv_xrc_conn_info xrc_info;
+	struct fi_ibv_xrc_cm_data cm_data;
+	const void *priv_data = cma_event->param.conn.private_data;
+	size_t priv_datalen = cma_event->param.conn.private_data_len;
+	int ret;
+
+	VERBS_DBG(FI_LOG_FABRIC, "EP %p INITIAL CONNECTION DONE state %d\n",
+		  ep, ep->conn_state);
+	ep->conn_state++;
+
+	/*
+	 * Original application initiated connect is done, if the passive
+	 * side of that connection initiate the reciprocal connection request
+	 * to create bidirectional connectivity.
+	 */
+	if (priv_data) {
+		ret = fi_ibv_eq_set_xrc_info(cma_event, &xrc_info);
+		if (ret) {
+			ep->conn_state--;
+			rdma_disconnect(ep->id);
+			goto err;
+		}
+		ep->peer_srqn = xrc_info.conn_data;
+		fi_ibv_ep_ini_conn_done(ep, xrc_info.conn_data,
+					xrc_info.conn_param.qp_num);
+		fi_ibv_eq_skip_xrc_cm_data(&priv_data, &priv_datalen);
+		fi_ibv_save_priv_data(ep, priv_data, priv_datalen);
+	} else {
+		fi_ibv_ep_tgt_conn_done(ep);
+		ret = fi_ibv_connect_xrc(ep, NULL, FI_IBV_RECIP_CONN, &cm_data,
+					 sizeof(cm_data));
+		if (ret) {
+			ep->conn_state--;
+			rdma_disconnect(ep->tgt_id);
+			goto err;
+		}
+	}
+err:
+	entry->info = NULL;
+	return -FI_EAGAIN;
```

Comment:
Again, why always FI_EAGAIN?

Reply:
In this case we are postponing the delivery of the FI_CONNECTED event, since this is only the data path connection from the initiator to the target. When the reciprocal connection is completed back to us we will return this connection event. So the FI_EAGAIN is returned because no event is delivered to the application and we want to keep the state machine moving forward.

Reply:
Should this be added as a comment?

Reply:
Yes, I'll add a comment here as well. Thanks.

Reply:
Done. Thanks.